services/director-v2/src/simcore_service_director_v2/modules/dask_client.py

```diff
@@ -48,28 +62,121 @@
 )
 
 logger = logging.getLogger(__name__)
-CLUSTER_RESOURCE_MOCK_USAGE: float = 1e-9
 
 
-def setup(app: FastAPI, settings: DaskSchedulerSettings) -> None:
-    async def on_startup() -> None:
-        await DaskClient.create(
-            app,
-            settings=settings,
+@dataclass
+class DaskSubSystem:
+    client: distributed.Client
+    gateway: Optional[dask_gateway.Gateway]
+    gateway_cluster: Optional[dask_gateway.GatewayCluster]
+
+    async def close(self):
+        if self.client:
+            await self.client.close()  # type: ignore
+        if self.gateway_cluster:
+            await self.gateway_cluster.close()  # type: ignore
+        if self.gateway:
+            await self.gateway.close()  # type: ignore
+
+
+async def _connect_to_dask_scheduler(endpoint: AnyUrl) -> DaskSubSystem:
+    try:
+        return DaskSubSystem(
+            await distributed.Client(
+                f"{endpoint}",
+                asynchronous=True,
+                name=f"director-v2_{socket.gethostname()}_{os.getpid()}",
+            ),
+            None,
+            None,
         )
+    except (TypeError) as exc:
+        raise ConfigurationError(
+            f"Scheduler has invalid configuration: {endpoint=}"
+        ) from exc
+
+
+DaskGatewayAuths = Union[
+    dask_gateway.BasicAuth, dask_gateway.KerberosAuth, dask_gateway.JupyterHubAuth
+]
+
+
+async def _get_gateway_auth_from_params(
+    auth_params: ClusterAuthentication,
+) -> DaskGatewayAuths:
+    try:
+        if isinstance(auth_params, SimpleAuthentication):
+            return dask_gateway.BasicAuth(**auth_params.dict(exclude={"type"}))
+        if isinstance(auth_params, KerberosAuthentication):
+            return dask_gateway.KerberosAuth()
+        if isinstance(auth_params, JupyterHubTokenAuthentication):
+            return dask_gateway.JupyterHubAuth(auth_params.api_token)
+    except (TypeError, ValueError) as exc:
+        raise ConfigurationError(
+            f"Cluster has invalid configuration: {auth_params}"
+        ) from exc
+    raise ConfigurationError(f"Cluster has invalid configuration: {auth_params=}")
+
+
+async def _connect_with_gateway_and_create_cluster(
+    endpoint: AnyUrl, auth_params: ClusterAuthentication
+) -> DaskSubSystem:
+    try:
+        gateway_auth = await _get_gateway_auth_from_params(auth_params)
+        gateway = dask_gateway.Gateway(
+            address=f"{endpoint}", auth=gateway_auth, asynchronous=True
+        )
+        # if there is already a cluster that means we can re-connect to it
+        cluster_reports_list = await gateway.list_clusters()
+        cluster = None
+        if cluster_reports_list:
+            cluster = await gateway.connect(
+                cluster_reports_list[0].name, shutdown_on_close=False
```

Comment:
Why take the first and not the 7th?

Reply:
well, there should be one and only one cluster on each gateway so far. that is why

Reply:
if that is assumed, I would state that in the code with 

```python
  if cluster_reports_list:
       assert len(cluster_reports_list)==1, "As of today, only one cluster per gateway"
       ...
```

Reply:
ok, done